File: 0874 Walking Robot Simulation/0874.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        x, y, z, ans = 0, 0, 0, 0
        invalid = set()

        for i, j in obstacles:
            invalid.add((i, j))

        for command in commands:
            if command < 0:
                z = (z + 1 if command == -1 else z + 3) % 4
            else:
                dx, dy = ((0, 1), (1, 0), (0, -1), (-1, 0))[z]
                while command > 0:
                    x, y = x + dx, y + dy
                    if (x, y) in invalid:
                        x, y = x - dx, y - dy
                        break
                    command -= 1
            ans = max(ans, x ** 2 + y ** 2)

        return ans

File: 0874 Walking Robot Simulation/0874.cpp
using namespace std;

class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        int x = 0, y = 0, z = 0, ans = 0;
        set<pair<int, int>> invalid;
        
        for (vector<int> obs: obstacles) invalid.insert({obs[0], obs[1]});

        for (int command: commands){
            if (command < 0) z = (z + (command == -1 ? 1 : 3)) % 4;
            else {
                vector<vector<int>> moves = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
                int dx = moves[z][0], dy = moves[z][1];
                while (command--){
                    x += dx; y += dy;
                    if (invalid.count({x, y})){
                        x -= dx; y -= dy;
                        break;
                    }
                }
                ans = max(ans, (int) (pow(x, 2) + pow(y, 2)));
            }
        }
        
        return ans;
    }
};
